### ISSUE TITLE
trying volta to work with neovim native lsp

### ISSUE THREAD

**Issue by thisguychris** [volta-cli/volta#813]
Hi, thanks for making volta! :wave: 

I'm trying to make volta global installs work with [neovim lsp-config](https://github.com/neovim/nvim-lspconfig#tsserver)

```
Error executing vim.schedule lua callback: 
....mount_nvimthatk4/usr/share/nvim/runtime/lua/vim
/lsp.lua:570: RPC[Error] code_name = InternalError, 
message = "Request initialize failed with message: Couldn't 
find 'tsserver' executable or 'tsserver.js' module"
```

the output of my `volta list` includes `tsserver`

```bash
 Tool binaries available:
        neovim-node-host (default)
        tsc, tsserver (default)
        typescript-language-server (default)
        vim-language-server (default)
```

If I use the native node's global install `npm install -g`. My setup works and LSP is able to call `tsserver`. I'm not sure what's the fundamental difference between the `volta install` and `npm install -g`.

But I did try with the flag `VOLTA_UNSAFE_GLOBAL=1 npm install -g typescript` for arbitrary install in global modules, but it still won't work. The only thing that works if I install native node and install via `npm install -g` command.

If I do install in the repo level, it finds `tsserver`

---
### Additional question:
Also, I would like to know if there's an additional overhead volta is adding when calling node binaries? Is it possible to have a long-running node, so that startup time would be faster when using for automatic formatting and linting?

**Comment by charlespierce** [comment]
Hi @thisguychris thanks for the questions! Given that `tsserver` is in your package list but it isn't found when run through `neovim`, the most likely cause is described in #492. Essentially, when we run a command, we currently remove the shim directory from the `PATH` (to prevent infinite recursion), which unfortunately results in each command / package being isolated an unable to call each other.

We're working right now on a full revamp of the global package install process, which will ultimately include a fix for this issue, allowing the packages to call one another.

The issue of `VOLTA_UNSAFE_GLOBAL=1 npm install -g typescript` not working also makes sense, as npm installs into the Volta directory where `node` is installed. That's actually part of why we (currently) disable global installs without the environment variable, they typically don't work so allowing them would add confusion. Another part of the package install rework we're doing will be to allow them to work correctly. If you use Yarn at all, I believe `VOLTA_UNSAFE_GLOBAL=1 yarn global add ...` tends to work better, because Yarn installs packages into a shared directory, instead of its own folder.

Finally, for your additional question, Volta does have a _small_ amount of overhead (one of our main project goals is to keep that as low as possible). On my MacBook Pro, that overhead is ~20ms, so it's tiny compared to the startup time of Node. As for whether it's possible to have a long-running Node, there's nothing in Volta that would prevent it, but it's not a feature we add. If there's some other tool that uses a long-running Node process, that should work with Volta, however I'm not sure if that's something Node itself supports.

**Comment by thisguychris** [comment]
@charlespierce Thank you for taking the time to answer all my questions. I only have an excitement about what comes next.

> We're working right now on a full revamp of the global package install process, which will ultimately include a fix for this issue, allowing the packages to call one another.

Is there a particular issue I should subscribe to?

> VOLTA_UNSAFE_GLOBAL=1 yarn global add ... tends to work better, because Yarn installs packages into a shared directory, instead of its own folder.

I'll try to test this, but just to give a background our team used to use yarn, but it was too green before and it contained lots of bugs. I'm sure it's more stable now. However, our team has moved back to npm since it fixed it's slowness issue, as well as adding automatic scanning. So this is more of a team decision for us. I understand you were just referring to the global package, however, installing yarn *would* be additional bulk for a tool we don't use exclusively, as well as may cause potential failures, accidentally committing yarn lock files, for example.

---
Sorry for the naive question. 
> If I use the native node's global install npm install -g. My setup works and LSP is able to call `tsserver`.

Just to add, Even if I uninstall native node, it successfully sees my global installs  `~/.npm-global/`. So I guess it correctly calls volta? Is it calling in the context of the `~/.npm-global` or under volta's internal packages, `~/.volta/tools...`?

**Comment by charlespierce** [comment]
> Is there a particular issue I should subscribe to?

The linked one #492 is probably the best. For context on all of the changes, you can check out the RFC here: https://github.com/volta-cli/rfcs/pull/45

> I'll try to test this, but just to give a background our team used to use yarn, but it was too green before and it contained lots of bugs. I'm sure it's more stable now. However, our team has moved back to npm since it fixed it's slowness issue, as well as adding automatic scanning. So this is more of a team decision for us. I understand you were just referring to the global package, however, installing yarn would be additional bulk for a tool we don't use exclusively, as well as may cause potential failures, accidentally committing yarn lock files, for example.

Totally reasonable, and if you aren't using Yarn already it's probably not the right choice to use it specifically for this one case.

> Just to add, Even if I uninstall native node, it successfully sees my global installs `~/.npm-global/`. So I guess it correctly calls volta? Is it calling in the context of the `~/.npm-global` or under volta's internal packages, `~/.volta/tools`...?

Interesting! I don't think `~/.npm-global` is a standard part of `npm` (unless it's something very new), so my guess is that however you are installing the native Node has some customizations to separate the global directory from the `npm` directory itself. Likely by setting a `prefix` configuration somewhere as part of the installation. Then those changes can persist even after Node is uninstalled. In that case what is likely happening is the script is being called in `~/.npm-global`, which is then executed using Volta's provided Node version. So it winds up being a bit of a hybrid.

**Comment by thisguychris** [comment]
Thanks, @charlespierce I'm rooting for volta! 

> Interesting! I don't think ~/.npm-global is a standard part of npm (unless it's something very new), so my guess is that however you are installing the native Node has some customizations to separate the global directory from the npm directory itself

Oh, you are totally right, I did this before: `npm config set prefix ~/.npm-global`
I find it better to install my global packages on a folder I own.

> So it winds up being a bit of a hybrid.

That's what I thought, so if it works now, I guess the only thing that's left is to make volta install like a regular node If you provided a custom prefix?

Anyway, again, thank you for being responsive to my questions. I've subscribed issue #492
